### PR TITLE
feature/starttls

### LINKED ISSUE #2406: configurable SMTP server
PacketFence currently only allow to use an external SMTP server without any credentials / security to relay outgoing emails. In the case of more "advanced" settings, a configuration of local Postfix is needed.

It would be great to have the ability to configure that kind of option directly within PacketFence.

screen shot 2017-06-16 at 10 11 35 am

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Add support for SSL and StartTLS for sending mail
Fixes #2406

Impacts
=======
Sending mail

NEW Package(s) required
=======================
- perl-IO-Socket-SSL-2.049-1.el7.centos.noarch (update)
- perl-libnet-3.10-1.1.noarch (update)
- perl-Socket-2.024-1.1.x86_64 (update)

Delete branch after merge
=========================
(REQUIRED)
NO

NEWS file entries
=================

Enhancements
------------
* Add support for SSL and StartTLS for sending mail

### CODE REVIEW

File: conf/documentation.conf [hunk 1]
```diff
@@ -648,6 +648,46 @@ violations with an action of "email", or any other PacketFence-related
...
+EOT
+
+[alerting.username]
+type=text
+description=<<EOT
+Username
+EOT
+
+[alerting.password]
+type=text
```
Comment (dwlfrth) [inline conf/documentation.conf]: That should be "password" type ?

File: lib/pf/config/util.pm [hunk 2]
```diff
@@ -474,6 +456,152 @@ sub get_captive_portal_uri {
...
+    eval {
+        $mime->send(
+            'sub',
+            \&send_by_pf_setting,
+            @args
+        );
+        $result = $mime->last_send_successful();
+    };
+    if ($@) {
+        my $msg = "Can't send email to $@";
```
Comment (dwlfrth) [inline lib/pf/config/util.pm]: Missing $email in log line

File: addons/packages/packetfence.spec [hunk 3]
```diff
@@ -139,6 +139,8 @@ Requires: ipset, sudo
 Requires: perl(File::Which), perl(NetAddr::IP)
 Requires: perl(Net::LDAP)
 Requires: perl(Net::IP)
+Requires: perl-libnet >= 3.10
+Requires: perl(Socket) >= 2.016
```
Comment (dwlfrth) [inline addons/packages/packetfence.spec:143]: missing perl-IO-Socket-SSL

**Comment by dwlfrth** [comment]
Check for packages in Debian

File: addons/packages/packetfence.spec [hunk 4]
```diff
@@ -139,6 +139,8 @@ Requires: ipset, sudo
 Requires: perl(File::Which), perl(NetAddr::IP)
 Requires: perl(Net::LDAP)
 Requires: perl(Net::IP)
+Requires: perl-libnet >= 3.10
```
Comment (fdurand) [inline addons/packages/packetfence.spec:142]: Does it work for debian too ?

File: conf/documentation.conf [hunk 5]
```diff
@@ -648,6 +648,46 @@ violations with an action of "email", or any other PacketFence-related
...
+description=<<EOT
+What encryption type to use.
+EOT
+
+[alerting.port]
+type=number
+description=<<EOT
+The port to connect to the SMTP.
+ If the port is set to 0 then port is calculated by the encryption type.
+ none : 25, ssl : 465, starttls : 587
```
Comment (julsemaan) [inline conf/documentation.conf]: no space between the encyrption type and the ':'

File: conf/pf.conf.defaults [hunk 6]
```diff
@@ -214,6 +214,33 @@ smtpserver=localhost
 #Subject prefix for email notifications of rogue DHCP servers, violations with an action of "email", or any other
 #PacketFence-related message.
 subjectprefix=PF Alert:
+#
+# alerting.encryption
+#
+encryption=none
+#
+# alerting.port
+#
```
Comment (julsemaan) [inline conf/pf.conf.defaults]: Missing the description

File: conf/pf.conf.defaults [hunk 7]
```diff
@@ -214,6 +214,33 @@ smtpserver=localhost
...
+#
+port=0
+#
+# alerting.username
+#
```
Comment (julsemaan) [inline conf/pf.conf.defaults]: Missing the description

File: conf/pf.conf.defaults [hunk 8]
```diff
@@ -214,6 +214,33 @@ smtpserver=localhost
 #Subject prefix for email notifications of rogue DHCP servers, violations with an action of "email", or any other
 #PacketFence-related message.
 subjectprefix=PF Alert:
+#
+# alerting.encryption
+#
+encryption=none
+#
```
Comment (julsemaan) [inline conf/pf.conf.defaults]: Missing the description

File: conf/pf.conf.defaults [hunk 9]
```diff
@@ -214,6 +214,33 @@ smtpserver=localhost
...
+# alerting.username
+#
+username=
+#
+# alerting.password
+#
+password=
+#
+# alerting.verify_ssl
+#
```
Comment (julsemaan) [inline conf/pf.conf.defaults]: Missing the description

File: conf/pf.conf.defaults [hunk 10]
```diff
@@ -214,6 +214,33 @@ smtpserver=localhost
...
+#
+verify_ssl=enabled
+#
+# alerting.timeout
+#
```
Comment (julsemaan) [inline conf/pf.conf.defaults]: Missing the description

File: lib/pf/config/util.pm [hunk 11]
```diff
@@ -474,6 +456,155 @@ sub get_captive_portal_uri {
...
+}
+
+=head2 send_by_pf_setting
+
+Handles the logic of sending an email of a MIME::Lite object
+Using the configuration of from pf.conf
+
+=cut
+
+sub send_by_pf_setting {
```
Comment (julsemaan) [inline lib/pf/config/util.pm]: send_using_pf_settings ?

File: lib/pfconfig/namespaces/config/Pf.pm [hunk 12]
```diff
@@ -133,8 +139,11 @@ sub build_child {
         $Config{omapi}{key_base64}=~ s/\R//g;   # getting rid of any carriage return
     }
 
-    return \%Config;
+    if ($Config{alerting}{port} == 0) {
+        $Config{alerting}{port} = $ALERTING_PORTS{$Config{alerting}{encryption}} // 25;
```
Comment (julsemaan) [inline lib/pfconfig/namespaces/config/Pf.pm]: Should have a constant for 25

**Comment by jrouzierinverse** [comment]
Done

File: conf/documentation.conf [hunk 13]
```diff
@@ -648,6 +648,46 @@ violations with an action of "email", or any other PacketFence-related
 message.
 EOT
 
+[alerting.encryption]
```
Comment (dwlfrth) [inline conf/documentation.conf]: Not quite sure about the configuration parameter name...
I'd say something like
alerting.smtp_encryption

File: conf/documentation.conf [hunk 14]
```diff
@@ -648,6 +648,46 @@ violations with an action of "email", or any other PacketFence-related
...
+type=toggle
+options=none|ssl|starttls
+description=<<EOT
+What encryption type to use.
+EOT
+
+[alerting.port]
```
Comment (dwlfrth) [inline conf/documentation.conf]: alerting.smtp_port ?

File: conf/documentation.conf [hunk 15]
```diff
@@ -648,6 +648,46 @@ violations with an action of "email", or any other PacketFence-related
...
+type=number
+description=<<EOT
+The port of the SMTP server.
+ If the port is set to 0 then port is calculated by the encryption type.
+ none: 25, ssl: 465, starttls: 587
+EOT
+
+[alerting.username]
```
Comment (dwlfrth) [inline conf/documentation.conf]: alerting.smtp_username ??

File: conf/documentation.conf [hunk 16]
```diff
@@ -648,6 +648,46 @@ violations with an action of "email", or any other PacketFence-related
...
+type=text
+description=<<EOT
+The username used to connect to the SMTP server
+EOT
+
+[alerting.password]
```
Comment (dwlfrth) [inline conf/documentation.conf]: alerting.smtp_password ?

File: conf/documentation.conf [hunk 17]
```diff
@@ -648,6 +648,46 @@ violations with an action of "email", or any other PacketFence-related
...
+type=obfuscated
+description=<<EOT
+The password used to connect to the SMTP server
+EOT
+
+[alerting.verify_ssl]
```
Comment (dwlfrth) [inline conf/documentation.conf]: alerting.smtp_verifyssl ?

File: conf/documentation.conf [hunk 18]
```diff
@@ -648,6 +648,46 @@ violations with an action of "email", or any other PacketFence-related
...
+type=toggle
+options=enabled|disabled
+description=<<EOT
+Verify SSL connection
+EOT
+
+[alerting.timeout]
```
Comment (dwlfrth) [inline conf/documentation.conf]: alerting.smtp_timeout ?

**Review by dwlfrth — COMMENTED** [review]
I'd say that outside of the configuration parameters name which I'm unsure of, that seems OK.
Also, have it running on a deployed setup and running fine.

**Comment by jrouzierinverse** [comment]
Done

**Review by dwlfrth — APPROVED** [review]
LGTM

**Pull request merged**